### ISSUE TITLE
Performance issue with producer

### ISSUE THREAD

**Issue by ctrix** [edenhill/librdkafka#1625]
Description
===========
I'm using rdkafka 0.11.1 (debian backported package) and i've used also the 0.9.3 version that is distributed with stretch. Kafka version is 1.0, freshly downloaded from the official package.

The test cluster is made of 3 Kafka VM and a producer that pushes messages to the brokers.
All VM are hosted on the same server and use KVM. Networking is 1GBit across all VM.

I'm pushing packets to a topic (3 partitions, replication set to 1) at the max possible speed.
The packets have a length from 50 to 400 bytes, uniformely distributed in that range.

What i'm experiencing is that i cannot use all the resources of the machine. The producer does not produce fast enough as it should.
In particular, the network usage is around 100Mbit, quite far from the 1Gbit cap limit.
CPUs (8 cores on the producer) are used at around 20%.

I tweaked quite a lot with all the producer config options but without success: more or less the results are always the same.
If i run two instances of the same program, i see that the bandwidth usage is doubled, meaning that the problem is not on the performance of the kafka cluster which is able to accept far more messages per second.

So i started testing using the example/rdkafka_producer (which i used as an example to write my code) to do the same and i hit the identical limits.

The only relevant parameter that makes a lot of difference is the size of the packets which, obviously, causes the network usage to grow (and saturate when it's too high).

So what i see is that the number of messages that both my code and the performance example can send to the kafka cluster is between 120K and 180K mps (the mileage may vary slightly when i use bigger message size, possibly because of the mallocs/frees).

Whatever i modify, i cannot increase significantly the number of messages that i can handle.

The fact that running two different programs effectively doubled the messages per second led me to try a different approach: I have run N (tried with 3,4,5,8) threads that receive from a nonblocking queue the messages.
Each thread creates a producer for the same topic and such thread produces messages into a different producer instance.

This approach failed miserabily, effectively lowering the performance.

Rough numbers:
my test tries to transfer 45 million messages. rdkafka sends all the messages in 280-320 seconds (compression changes something). If i comment the calls to produce() the tests exits in 150seconds.

So, what is the problem here ?
Is there a way i can increase the performance and hit higher messages per second ?

Checklist
=========
Please provide the following information:

 - [x] librdkafka version (release number or git tag): `0.11.1` and 0.9.3, and git tag b1182718dbd8a0cb7a21527739552479711078e1 (Dec. 23)
 - [x] Apache Kafka version: `1.0.0`
 - [x] Operating system: `Linux master 4.9.0-4-amd64 #1 SMP Debian 4.9.65-3+deb9u1 (2017-12-23) x86_64 GNU/Linux`
 - [x] librdkafka client configuration: as follows

  3) message.max.bytes                        1000000
  4) message.copy.max.bytes                   65535
  5) receive.message.max.bytes                100000000
  6) max.in.flight.requests.per.connection    10000
  7) metadata.request.timeout.ms              60000
  8) topic.metadata.refresh.interval.ms       300000
  9) metadata.max.age.ms                      -1
 10) topic.metadata.refresh.fast.interval.ms  250
 11) topic.metadata.refresh.fast.cnt          10
 12) topic.metadata.refresh.sparse            true
 14) socket.timeout.ms                        60000
 15) socket.blocking.max.ms                   1000
 16) socket.send.buffer.bytes                 0
 17) socket.receive.buffer.bytes              0
 18) socket.keepalive.enable                  false
 19) socket.nagle.disable                     false
 20) socket.max.fails                         3
 21) broker.address.ttl                       1000
 22) broker.address.family                    any
 23) reconnect.backoff.jitter.ms              500
 24) statistics.interval.ms                   0
 25) enabled_events                           0
 26) log_cb                                   0x7f3441cec0a0
 27) log_level                                6
 28) log.queue                                false
 29) log.thread.name                          true
 30) log.connection.close                     true
 31) socket_cb                                0x7f3441cff970
 32) open_cb                                  0x7f3441d10f10
 33) internal.termination.signal              29
 34) api.version.request                      true
 35) api.version.request.timeout.ms           10000
 36) api.version.fallback.ms                  1200000
 37) broker.version.fallback                  0.9.0
 44) partition.assignment.strategy            range,roundrobin
 45) session.timeout.ms                       30000
 46) heartbeat.interval.ms                    1000
 47) group.protocol.type                      consumer
 48) coordinator.query.interval.ms            600000
 49) enable.auto.commit                       true
 50) auto.commit.interval.ms                  5000
 51) enable.auto.offset.store                 true
 52) queued.min.messages                      100000
 53) queued.max.messages.kbytes               1000000
 54) fetch.wait.max.ms                        100
 55) fetch.message.max.bytes                  1048576
 56) fetch.min.bytes                          1
 57) fetch.error.backoff.ms                   500
 58) offset.store.method                      broker
 59) enable.partition.eof                     true
 60) check.crcs                               false
 61) queue.buffering.max.messages             1000000
 62) queue.buffering.max.kbytes               4000000
 63) queue.buffering.max.ms                   500
 64) message.send.max.retries                 2
 65) retry.backoff.ms                         100
 66) compression.codec                        lz4
 67) batch.num.messages                       1000000
 68) delivery.report.only.error               false

**Comment by tinux2015** [comment]
queue.buffering.max.ms 500
Do you try to increase this config ?

**Comment by edenhill** [comment]
> So i started testing using the example/rdkafka_producer 

I suppose you mean `examples/rdkafka_performance` (in producer mode)?

The main thing that affect throughput is batch sizes; the larger the batches the higher the throughput.
To get a feel a for what your batching situation looks like I suggest you enable `debug=msg`, which is going to be noisy but valuable.
Look for lines like:
`MessageSet with XX message(s)..`
What typical number of messages do you get? Are all successfully delivered or does it report failures?

You can also enable per-request debugging with `debug=msg,protocol` and look for `ProduceResponse .. rtt YY` printouts, that will tell us how much time each Produce request takes full network cycle.

**Comment by ctrix** [comment]
Thanks for the help.
Sorry for the fault, yes, i was intending the rdkafka_performance example.

I've tweaked the batch.num.messages parameter trying to understand which batch size would fit better. it seems that lowering it indeed makes things slower but an increased value does not improve much.

I'm attaching a small report indicating all the common options in all the runs, changing only batch.num.messages.

what i see is that the max number of messages that is sent to the server for each batch is around 6000, given that the batch size is large enough.

As suggested, i also increased queue.buffering.max.ms to 1000 and higher values but i didn't get any performance gain.

I think that the problem may depend on this:
%7|1515288216.914|PRODUCE|rdkafka#producer-1| [thrd:pdns-kafka-3:9092/3]: pdns-kafka-3:9092/3: No more space in current MessageSet (6181 message(s), 999921 bytes)

And in particular on the 999921 (too much close to 1000000). What is the parameter that sets this limit ?
I tried with queue.buffering.max.kbytes and queued.max.messages.kbytes with no luck.

[report.txt](https://github.com/edenhill/librdkafka/files/1609412/report.txt)
